### PR TITLE
Unify bits, arch, and android_arch into env["arch"]

### PR DESCRIPTION
**Pull request by aaronfranke**
This PR addresses part of https://github.com/godotengine/godot-proposals/issues/1416

In the master branch before this PR, we have several arguments for specifying the architecture: `bits`, `arch`, and `android_arch`. This is a bit of a mess, so this PR replaces all three of these with a single unified `arch`. Note that `env["bits"]` still exists internally for now (or it could be kept forever).

The list of valid architectures is `["auto", "x86_32", "x86_64", "arm32", "arm64", "rv64", "ppc32", "ppc64", "wasm32"]`. The names were chosen to be explicit such that each architecture always includes the bitness in the name (except for "auto").

If `"auto"`, then it will automatically determine the architecture. It will use arm64 for Android and iOS, wasm32 for JavaScript, and otherwise it will use the host architecture, so arm64 on modern ARM devices, x86_64 on a typical modern desktop, and x86_32 on an old 32-bit desktop.

For macOS, iOS, JavaScript, and Android, it will throw an error if trying to build for an invalid architecture. The same should be done for Windows and Linux but I'm not sure how to get cross-compiling working in all cases where it should exist or how to show an error when the necessary libs etc are missing. There's also several places in both Windows and Linux that are hard-coded for x86. I left some TODOs in the Windows and Linux code for future PRs.

I've tested this PR compiling on an arm64 macOS host for x86_64 and arm64 macOS and for arm64 iOS, on an arm64 Ubuntu VM for arm64 Linux, on a rv64 Ubuntu VM for rv64 Linux, on an x86_64 Ubuntu host for x86_64 Linux, on an x86_32 Ubuntu VM for x86_32 Linux (fails linking, but happens on master too), on an x86_64 Windows host for x86_32 and x86_64 Windows using MSVC and x86_64 using MinGW, on both an x86_64 Windows and x86_64 Ubuntu host for x86_32, x86_64, arm32, and arm64 Android (x86 Android fails linking on Ubuntu but this happens on master too, and I only tested if the arm64 Android build runs), and on an x86_64 Windows host for wasm32 JavaScript. Still, it needs further testing. Most of the platforms I've tested on have revealed bugs (some of which exist on master too).

This PR also simplifies a lot of architecture checks (such as the ones in the denoise module).

@q66 Can you test if this works as expected on PowerPC?

### CODE REVIEW

File: platform/iphone/detect.py [hunk 1]
```diff
@@ -43,6 +43,13 @@ def get_flags():
 
 def configure(env):
     ## Build type
+    if env["arch"] != "arm64":
+        # iOS no longer runs 32-bit apps as of iOS 11 which was
+        # released in 2017. As such, we only support 64-bit.
+        print("Unsupported CPU architecture: " + env["arch"] + ". Only arm64 is supported on iOS.")
```
Comment (bruvzg) [inline platform/iphone/detect.py:49]: `x86_64` simulator for Intel macs still exists.

File: modules/mono/build_scripts/mono_configure.py [hunk 2]
```diff
@@ -406,7 +406,7 @@ def make_template_dir(env, mono_root):
 
     assert is_desktop(platform)
 
-    template_dir_name = "data.mono.%s.%s.%s" % (platform, env["bits"], target)
+    template_dir_name = "data.mono.%s.%s.%s" % (platform, env["arch"], target)
```
Comment (neikeq) [inline modules/mono/build_scripts/mono_configure.py:409]: This will break C# exporter code in `GodotTools`, which is unchanged in this PR.

**Review by neikeq — CHANGES_REQUESTED** [review]

**Comment by neikeq** [comment]
I think the drop of 32bit support on iOS should go into its own commit.
It's much better to have `Drop iOS 32-bit support` commit than for these changes to be hidden in an unrelated commit.

File: modules/mono/build_scripts/mono_configure.py (hunk 2 above)
Comment (neikeq) [reply modules/mono/build_scripts/mono_configure.py:409]: This are the lines:
https://github.com/godotengine/godot/blob/21b81998bb42b2324a539f01354b68e4fba2c696/modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs#L334-L338

**Comment by akien-mga** [comment]
> I think the drop of 32bit support on iOS should go into its own commit. It's much better to have `Drop iOS 32-bit support` commit than for these changes to be hidden in an unrelated commit.

Yes, and similarly I would put the actual build fixes in a dedicated PR too. This should only focus on buildsystem changes.
(In the same vein, I didn't comment as I didn't want to delay further, but for #55572 I would also have preferred separate PRs for actually fixing the `double` build and adding it to the CI.)

**Comment by aaronfranke** [comment]
Opened #55792 and #55792 which take pieces out of this PR.

File: modules/mono/build_scripts/mono_configure.py (hunk 2 above)
Comment (aaronfranke) [reply modules/mono/build_scripts/mono_configure.py:409]: I looked around and this would require a lot of work to fix, as it would involve adding a dropdown in ExportPlugin for the architecture so that this information can be passed down to here. I'll see what I can do.

File: SConstruct [hunk 3]
```diff
@@ -119,13 +120,29 @@ opts.Add("p", "Platform (alias for 'platform')", "")
...
 opts.Add(EnumVariable("target", "Compilation target", "debug", ("debug", "release_debug", "release")))
-opts.Add("arch", "Platform-dependent architecture (arm/arm64/x86/x64/mips/...)", "")
-opts.Add(EnumVariable("bits", "Target platform bits", "default", ("default", "32", "64")))
 opts.Add(EnumVariable("float", "Floating-point precision", "default", ("default", "32", "64")))
 opts.Add(EnumVariable("optimize", "Optimization type", "speed", ("speed", "size", "none")))
 opts.Add(BoolVariable("production", "Set defaults to build Godot for use in production", False))
 opts.Add(BoolVariable("use_lto", "Use link-time optimization", False))
 
+# CPU architecture options.
+architectures = ["", "x86_32", "x86_64", "arm32", "arm64", "rv64", "ppc32", "ppc64", "wasm32"]
```
Comment (m4gr3d) [inline SConstruct:129]: For `x86` and `arm` architectures, do we really need to change their nomenclature within the build configs?

Especially as `x86` usually indicates 32 bits, while `x86_64` indicates 64 bits. Similarly for `arm`, `armeabi-v7a` indicates 32 bits and `arm64-v8a` indicates 64 bits.

Seems like we can just use the same terms as used by everyone else instead of coming up with our own (`arm32`, `arm64`) terms.

File: platform/android/java/lib/build.gradle [hunk 4]
```diff
@@ -105,7 +105,7 @@ android {
         def taskName = getSconsTaskName(buildType)
         tasks.create(name: taskName, type: Exec) {
             executable sconsExecutableFile.absolutePath
-            args "--directory=${pathToRootDir}", "platform=android", "target=${releaseTarget}", "android_arch=${defaultAbi}", "-j" + Runtime.runtime.availableProcessors()
+            args "--directory=${pathToRootDir}", "platform=android", "target=${releaseTarget}", "arch=${defaultAbi}", "-j" + Runtime.runtime.availableProcessors()
```
Comment (m4gr3d) [inline platform/android/java/lib/build.gradle:108]: You'll also need to update the [`supportedAbis`](https://github.com/godotengine/godot/blob/15aea89868d16d0310a229d831e4a347a38dc1c3/platform/android/java/build.gradle#L24) values from which `defaultAbi` is selected.

File: SConstruct (hunk 3 above)
Comment (aaronfranke) [reply SConstruct:129]: Yes, we should be explicit. 32-bit versions of these architectures are getting to be more and more niche and it's common to refer to both 32-bit and 64-bit versions of x86 as "x86". We want to avoid people wanting to build "for x86" or "for ARM" and accidentally building for what-will-be-ancient 32-bit versions of them. Also, these terms are not dominant but they are not unique to Godot either, I've seen many usages of "x86_32" online.

File: modules/mono/build_scripts/mono_configure.py (hunk 2 above)
Comment (neikeq) [reply modules/mono/build_scripts/mono_configure.py]: Are you sure about that? I think features includes the architectures. See:

https://github.com/godotengine/godot/blob/21b81998bb42b2324a539f01354b68e4fba2c696/modules/mono/editor/GodotTools/GodotTools/Export/AotBuilder.cs#L536-L557

Or is that only added for mobile platforms?

Alternatively, since features still includes the bits (this PR doesn't seem to remove that), you could change that to `features.Contains("64") ? "x86_64" : "x86_32"`.

Comment (aaronfranke) [reply modules/mono/build_scripts/mono_configure.py]: I think we'd want to make this work for any target architecture, so I'm inclined on passing `env["arch"]` all the way down. However it could also be done in a follow-up PR, so in the meantime I changed the above back to `env["bits"]`.

**Review by neikeq — APPROVED** [review]

File: platform/osx/export/export_plugin.cpp [hunk 5]
```diff
@@ -628,19 +628,19 @@ Error EditorExportPlatformOSX::export_project(const Ref<EditorExportPreset> &p_p
...
+				file = file.replace("/data.mono.osx.x86_64.release_debug/", "/GodotSharp/");
 			}
-			if (file.find("/data.mono.osx.64.release/") != -1) {
+			if (file.find("/data.mono.osx.x86_64.release/") != -1) {
 				if (p_debug) {
 					ret = unzGoToNextFile(src_pkg_zip);
 					continue; // skip
 				}
-				file = file.replace("/data.mono.osx.64.release/", "/GodotSharp/");
+				file = file.replace("/data.mono.osx.x86_64.release/", "/GodotSharp/");
```
Comment (bruvzg) [inline platform/osx/export/export_plugin.cpp:643]: This should stay `64` or skip architecture at all, usually this path point to the Universal (`x86_64` + `arm64`) binaries.

Comment (aaronfranke) [reply platform/osx/export/export_plugin.cpp]: Since this is supposed to point to universal, I reverted this part for this PR, but I think it would make sense to remove the architecture in the future for universal things.

**Comment by aaronfranke** [comment]
I tested the Android builds using these commands and it succeeded:

```bash
scons werror=yes -j8 platform=android arch=arm32
scons werror=yes -j8 platform=android arch=arm64
scons werror=yes -j8 platform=android arch=x86_32
scons werror=yes -j8 platform=android arch=x86_64

cd platform/android/java
./gradlew generateGodotTemplates
```

Note that I haven't tested both Android and Mono at the same time as I haven't been able to get it working. Today I've tested these Android architectures building from ARM macOS and I've tested Mono on ARM Linux.

File: platform/osx/export/export_plugin.cpp (hunk 5 above)
Comment (akien-mga) [reply platform/osx/export/export_plugin.cpp]: Yeah I think it can be left out for this PR, but that's something that needs more work later.

We can indeed remove the bits/arch part which is currently irrelevant. But we have to think about whether we want to actually make it possibility to support different architectures with this like we do on other platforms, and thus have:
```
data.mono.osx.x86_64.release
data.mono.osx.arm64.release
data.mono.osx.universal.release
```

Official builds would use the latter, but users who want to ship separate x86_64 and arm64 builds for some reason would still have the option to do it with pre-defined names. Currently you'd have to use `data.mono.osx.64.release` regardless of what the actual native binaries inside are compiled for.

**Comment by q66** [comment]
@aaronfranke this cleanup looks very good in concept and answers some distribution concerns for me, so thanks for that; I can't test it on ppc right now, but I don't see anything that would break it. I will fix any possible fallout with regard to the platform at a later time.

File: platform/android/SCsub [hunk 6]
```diff
@@ -55,7 +55,7 @@ if lib_arch_dir != "":
     else:  # release_debug, debug
         lib_type_dir = "debug"
 
-    out_dir = "#platform/android/java/lib/libs/" + lib_type_dir + "/" + lib_arch_dir
+    out_dir = "#platform/android/java/lib/libs/" + lib_type_dir + "/" + env["arch"]
```
Comment (m4gr3d) [inline platform/android/SCsub:58]: This should be `lib_arch_dir` instead of `env["arch"]` as the Android build system will look specifically for these ABIs terms.

File: platform/android/java/app/config.gradle [hunk 7]
```diff
@@ -127,7 +127,7 @@ final String VALUE_SEPARATOR_REGEX = "\\|"
 ext.getExportEnabledABIs = { ->
     String enabledABIs = project.hasProperty("export_enabled_abis") ? project.property("export_enabled_abis") : "";
     if (enabledABIs == null || enabledABIs.isEmpty()) {
-        enabledABIs = "armeabi-v7a|arm64-v8a|x86|x86_64|"
+        enabledABIs = "arm32|arm64|x86_32|x86_64|"
```
Comment (m4gr3d) [inline platform/android/java/app/config.gradle:130]: This is not going to work as the Android build system expects specific terms for the ABIs it supports (https://developer.android.com/ndk/guides/abis#gc).

**Review by m4gr3d — DISMISSED** [review]
On further review, I'm realizing there are effects of this PR that may be overlooked during review (see added comments as examples). 
So validating it works as expected (and doesn't introduce regressions) will require full manual build/test/evaluation phase with actual projects.

File: platform/android/SCsub (hunk 6 above)
Comment (aaronfranke) [reply platform/android/SCsub:58]: With the changes I made, this line expects `arm32` etc. Unless there are other places that are also incorrect. It's hard to tell specifically what is incorrect since it does compile fine (but just changing this one line to `lib_arch_dir` doesn't build if I recall correctly, I can retest later but I'm not available at the moment, maybe I just need to change what's in your other comment).

I do remember for sure that Line 64 does expect the values in `lib_arch_dir` as that points to sources in the NDK.

Comment (m4gr3d) [reply platform/android/SCsub:58]: It compiles fine, but I expect it to crash at runtime as the system will be unable to locate and load the relevant shared libraries.

Comment (aaronfranke) [reply platform/android/SCsub]: I changed this code and it builds fine again. Note that the export dialog doesn't work yet. That's a lot of work on its own so I'm considering either making that a second commit on this PR or making that be a follow-up PR.

**Comment by aaronfranke** [comment]
I updated this PR to add `arch=universal` on macOS, which I learned how to do thanks to @bruvzg's comment here https://github.com/godotengine/godot-cpp/pull/714#discussion_r815596024 (but I haven't figured out how to do this on iOS for the engine yet). This means that this PR is no longer purely refactoring, it also adds the ability to build a universal binary on macOS.

I also did some work on the export dialog on another branch. There is some other refactoring that should be done in the export system, such as how EditorExportPlatformPC is currently used for LinuxBSD but there should really be a separate class for LinuxBSD. It's a decent chunk of work, it might make sense to do as a different PR, although if we merge this work without the export dialog changes then you would be temporarily unable to export things in master.

**Comment by nathanfranke** [comment]
Testing out GDExtension exports, looks like this PR resolves the feature flags for android (`arm64-v8a` was correctly changed to `arm64`), but not quite for Linux. The feature list has `64`, while my GDExtension libraries and files use `x86_64`.

File: platform/linuxbsd/detect.py [hunk 8]
```diff
@@ -79,26 +79,20 @@ def configure(env):
...
+    elif env["arch"] == "x86_32":
+        env.Append(CCFLAGS=["-m32", "-march=i686"])
+        env.Append(LINKFLAGS=["-m32", "-march=i686"])
+    elif env["arch"] == "arm64":
+        env.Append(CCFLAGS=["-march=armv8-a"])
+        env.Append(LINKFLAGS=["-march=armv8-a"])
+    elif env["arch"] == "rv64":
         # G = General-purpose extensions, C = Compression extension (very common).
         env.Append(CCFLAGS=["-march=rv64gc"])
+        env.Append(LINKFLAGS=["-march=rv64gc"])
```
Comment (akien-mga) [inline platform/linuxbsd/detect.py:95]: IMO that's scope creep and this should be left for another PR.

This one is big / regression prone enough that we shouldn't add new features at the same time that we're refactoring.

File: platform/android/export/export_plugin.cpp [hunk 9]
```diff
@@ -590,9 +590,9 @@ zip_fileinfo EditorExportPlatformAndroid::get_zip_fileinfo() {
 
 Vector<String> EditorExportPlatformAndroid::get_abis() {
 	Vector<String> abis;
-	abis.push_back("armeabi-v7a");
-	abis.push_back("arm64-v8a");
-	abis.push_back("x86");
+	abis.push_back("arm32");
+	abis.push_back("arm64");
+	abis.push_back("x86_32");
 	abis.push_back("x86_64");
```
Comment (akien-mga) [inline platform/android/export/export_plugin.cpp:596]: I'm not sure this is correct to assume that `abis` == `env["arch"]` in this context.

The previous values were matching what's actually used by Android toolchains and if those are used as configuration values for gradle it might break exports.

CC @m4gr3d

File: modules/mono/editor/GodotTools/GodotTools/Export/AotBuilder.cs [hunk 10]
```diff
@@ -547,9 +547,9 @@ private static IEnumerable<string> GetEnabledAndroidAbis(string[] features)
         {
             var androidAbis = new[]
             {
-                "armeabi-v7a",
-                "arm64-v8a",
-                "x86",
+                "arm32",
+                "arm64",
+                "x86_32",
```
Comment (akien-mga) [inline modules/mono/editor/GodotTools/GodotTools/Export/AotBuilder.cs:552]: This is probably wrong. It doesn't match the changes you made in `mono_configure.py`, as you can see those were the values from the dict, not the keys.

**Comment by akien-mga** [comment]
There are quite a few things I wanted to refactor in the proposed implementation, and instead of making a ton of review comments with vague suggestions I preferred to just do the work myself -- my changes are of course up for review/discussion too.

I pushed my changes directly to your branch as a separate commit so they can be assessed, and we can keep reviewing the whole changes from the PR in parallel.

The main things I changed are:

- Dehardcoded the default arch handling from `SConstruct`. Platform-specific should stay in the platform `detect.py` as far as possible. Thankfully `get_flags()` lets us define a default value easily for each platform, so it works out of the box when not specified. One caveat I haven't address yet is that manually running `scons arch=auto` fails as it then overrides the auto-detection from `get_flags()` with an invalid value.
  * Moved the list of allowed arches and the `platform.machine()` detection to `platform_methods` so they can be used in `platform/*/detect.py`, and that de-clutters `SConstruct`.
  * Added `x86` alias for `x86_32` too.
- Nearly fully removed the `env["bits"]` hacks.
  * Mono still uses "32" and "64" for some stuff so I made the check ad hoc.
  * UWP, and MSVC "manual" config are probably broken and will need further work.
- Unified printing what platform, tools, target and arch we build for.
- Unified code to check that the selected arch is valid for the selected platform.
- Removed `arch=universal` because it makes it harder to know what actual architecture(s) we're targeting from other code locations. Having to do platform specific checks is pretty much the main issue that motivated this change, so we shouldn't re-add configuration options that make things uncertain. Notably we need to be able to know if we're building for `x86_64` only, or `arm64` only (and when checking for both users would need to remember to check also `universal`). For example `universal` would break this logic: `if env["builtin_libtheora"] and env["arch"] == "x86_64":`

**Comment by aaronfranke** [comment]
Excellent work, thank you!

> For example `universal` would break this logic: `if env["builtin_libtheora"] and env["arch"] == "x86_64":`

My thinking here was that if some optional feature only works for one of the two architectures in universal, then we should include it in neither for consistency, to avoid a feature working on one architecture and being missing on another. But since this is about optimizations, then I guess the only option if you want those is to build x86_64 and arm64 separately and combine them later.

File: methods.py [hunk 11]
```diff
@@ -777,7 +777,8 @@ def __getitem__(self, k: str):
         if env.get("module_mono_enabled"):
             import modules.mono.build_scripts.mono_reg_utils as mono_reg
 
-            mono_root = env.get("mono_prefix") or mono_reg.find_mono_root_dir(env["bits"])
+            bits = "32" if env["arch"] in ["arm32", "x86_32"] else "64"
```
Comment (aaronfranke) [inline methods.py:780]: Another option is `bits = "32" if "32" in env["arch"] else "64"`, it may seem more hacky but it should always work since it would be weird for an architecture to contain `32` without it being 32-bit.

**Review by aaronfranke — COMMENTED** [review]
The `detect_arch` method and arch detection refactoring makes sense, since it splits platform-specific stuff out. However, we do still need a way to handle `arch=auto` as you mentioned. I don't know if there's any better way other than having a very specific `if` statement for it.

**Comment by aaronfranke** [comment]
I updated this PR to handle the case of `arch=auto`.

Based on Akien's comments above, I opened another PR https://github.com/godotengine/godot/pull/64366 which is mostly a subset of this PR that only changes the Linux build system logic, and I marked this one as a draft since we can review and merge the smaller PR first.

**Comment by akien-mga** [comment]
I would remove the `-march` changes from this PR, it's not needed to merge `arch` option refactoring.

File: platform/windows/detect.py [hunk 12]
```diff
@@ -77,11 +79,14 @@ def get_opts():
...
+    ]
 
 
 def build_res_file(target, source, env):
-    if env["bits"] == "32":
+    # TODO: This shouldn't be hard-coded for x86.
+    if env["arch"] == "x86_32":
         cmdbase = env["mingw_prefix_32"]
     else:
         cmdbase = env["mingw_prefix_64"]
```
Comment (bruvzg) [inline platform/windows/detect.py:92]: I guess it should be simply a single `mingw_prefix` that can be set alongside `arch`.

File: platform/windows/detect.py [hunk 13]
```diff
@@ -100,20 +105,28 @@ def build_res_file(target, source, env):
...
 
-    # Force bits arg
+    # Force ARCH arg
     # (Actually msys2 mingw can support 64-bit, we could detect that)
-    env["bits"] = "32"
+    # TODO: This is wrong, but not sure what to do about it.
+    # We want to determine the arch and bitness in the SConstruct only.
+    # We can check if it's correct in here, but if it's not, it should
+    # just fail with an error message instead of trying to force it.
+    env["arch"] = "x86_32"
```
Comment (bruvzg) [inline platform/windows/detect.py:129]: This should be fine, this part of code is used when you are running scons from a developer console for a specific architecture. Architecture is detected by `detect_visual_c_compiler_version`(which is looking in the `PATH` to get compiler name). But I would move `env["arch"] = "x86_32"` to the `elif compiler_version_str == "x86" ...` branch below.

File: platform/windows/detect.py [hunk 14]
```diff
@@ -122,15 +135,15 @@ def setup_msvc_manual(env):
...
-        print("Compiled program architecture will be a 32 bit executable. (forcing bits=32).")
+        print("Compiled program architecture will be a 32 bit executable (forcing arch=x86_32).")
     else:
         print(
-            "Failed to manually detect MSVC compiler architecture version... Defaulting to 32bit executable settings"
-            " (forcing bits=32). Compilation attempt will continue, but SCons can not detect for what architecture this"
+            "Failed to manually detect MSVC compiler architecture version... Defaulting to x86 32-bit executable settings"
+            " (forcing arch=x86_32). Compilation attempt will continue, but SCons can not detect for what architecture this"
             " build is compiled for. You should check your settings/compilation setup, or avoid setting VCINSTALLDIR."
         )
```
Comment (bruvzg) [inline platform/windows/detect.py:148]: This should just fail, there's no point to continue if compiler is not present.

File: platform/windows/detect.py [hunk 15]
```diff
@@ -328,7 +346,7 @@ def configure_mingw(env):
         env.Append(CCFLAGS=["-msse2"])
 
         if env["optimize"] == "speed":  # optimize for speed (default)
-            if env["bits"] == "64":
+            if env["arch"] == "x86_64":
                 env.Append(CCFLAGS=["-O3"])
             else:
                 env.Append(CCFLAGS=["-O2"])
```
Comment (bruvzg) [inline platform/windows/detect.py:352]: That's probably were added specifically for `x86`, and should be:

```
            if env["arch"] == "x86_32":
                env.Append(CCFLAGS=["-O2"])
            else:
                env.Append(CCFLAGS=["-O3"])
```

File: platform/windows/detect.py [hunk 16]
```diff
@@ -365,15 +383,12 @@ def configure_mingw(env):
...
-            env["bits"] = "64" if "PROGRAMFILES(X86)" in os.environ else "32"
-        else:  # default to 64-bit on Linux
-            env["bits"] = "64"
-
     mingw_prefix = ""
 
-    if env["bits"] == "32":
+    # TODO: This doesn't seem to be working, or maybe I just have
+    # MinGW set up incorrectly. It always gives me x86_64 builds,
+    # even though arch == "x86_32" and the file name has x86_32 in it.
```
Comment (bruvzg) [inline platform/windows/detect.py:390]: Depending on setup, MinGW can have either prefixed binaries (e.g. `x86_64-w64-mingw32-gcc`) or everything in the subfolder (`gcc`) in this case which one used is determined by `path`. I'm not sure how to detect arch in this case, we can do something like `detect_visual_c_compiler_version`, but MinGW path is not standard.

**Review by bruvzg — APPROVED** [review]
Using `-march=` seems to work fine with MinGW, but `x86-64` for example only available if running from 64-bit MSYS2 environment.

So I guess the best way is to, try to set prefix based on `arch` if it's not set (ignore it otherwise since it can be something line `/mingw/bin/` without arch name). Always set `-march` to make sure it's building for the correct one. Plus check `MSYS2` env variables and warn if trying to use 64-bit arch in 32-bit env.

For MSVC, I think we can always force compilers path based on arch, even if it's started from wrong environment (maybe with a warning as well).

In any case, this should be good to merge as is (after rebase), and I'll add both checks / `-march` for Windows cross compile when implementing ARM support for Windows.

File: modules/mono/build_scripts/mono_configure.py [hunk 17]
```diff
@@ -128,27 +128,23 @@ def get_runtime_path():
 
 
 def determine_runtime_identifier(env):
+    # The keys are Godot's names, the values are the Microsoft's names.
+    # List: https://docs.microsoft.com/en-us/dotnet/core/rid-catalog
     names_map = {
         "windows": "win",
         "macos": "osx",
         "linuxbsd": "linux",
         "server": "linux",  # FIXME: Is server linux only, or also macos?
```
Comment (neikeq) [inline modules/mono/build_scripts/mono_configure.py:137]: I thought the server platform was removed.

Comment (aaronfranke) [reply modules/mono/build_scripts/mono_configure.py:137]: It is removed from the engine, and will be removed here when https://github.com/godotengine/godot/pull/64846 is merged.

File: platform/windows/detect.py (hunk 14 above)
Comment (akien-mga) [reply platform/windows/detect.py:148]: The print is no longer accurate then, since it says explicitly that it's falling back to x86 32-bit settings.

File: platform/windows/detect.py (hunk 15 above)
Comment (akien-mga) [reply platform/windows/detect.py]: Do we know why `x86_32` would need to use `-O2` instead of `-O3`?
Can there be similar issues on `arm32`?

**Comment by akien-mga** [comment]
I'm pretty happy with the state of this PR now, I'll do a final review and squash the commits so we can merge.

**Comment by akien-mga** [comment]
I reviewed the changes again and amended a couple nitpicks.

I tested an Android build with:
```
scons p=android arch=arm64 tools=no target=release_debug
cd platform/android/java
./gradlew generateGodotTemplates
```

And it seems to work fine. I could use the resulting APK and source templates to do both APK and Custom Build/AAB exports to my arm64 Poco F1, which ran fine.

I still have some concerns about the Android+Mono changes, and I think some renaming might be in order to make a clear separation between the `arch` as used by Godot SCons and the `abi` as expected by Gradle/Android. We should avoid assigning Godot's `env["arch"]` names to variables named `abi` IMO. But I think this can be cleaned up in future PRs together with testing and if needed fixing the Android+Mono export code.

**Review by akien-mga — APPROVED** [review]
Amazing work!

**Comment by akien-mga** [comment]
Thanks!

File: platform/android/export/export_plugin.cpp (hunk 9 above)
Comment (m4gr3d) [reply platform/android/export/export_plugin.cpp:594]: @akien-mga @aaronfranke This section is indeed incorrect as those values are passed to gradle build process which expects the Android supported abis.

Comment (Zireael07) [reply platform/android/export/export_plugin.cpp:594]: Worth a comment likely?

Comment (m4gr3d) [reply platform/android/export/export_plugin.cpp:594]: @Zireael07 I've synced with @akien-mga and @aaronfranke and working on a fix for it.

File: platform/osx/export/export_plugin.cpp (hunk 5 above)
Comment (aaronfranke) [reply platform/osx/export/export_plugin.cpp]: For the record the names were changed here: https://github.com/godotengine/godot/pull/63096/files#diff-72b5d6bdf8391a9477aca91fe06263335d88623f353c71cc56ffc8bda6102221
